### PR TITLE
DOC-103: MultiPart impl for saving

### PR DESCRIPTION
**Pull request by yatsenko-ihor**

### CODE REVIEW

File: datasafe-storage/datasafe-storage-impl-s3/src/test/java/de/adorsys/datasafe/business/impl/storage/MultipartUploadS3SystemStorageServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+        String testFileChecksum = calculateChecksumOfTestFile(testFileName);
+
+        AbsoluteLocation<PrivateResource> privateLocation = new AbsoluteLocation<>(
+                BasePrivateResource.forPrivate(URI.create("s3://" + bucketName + "/bigfile.txt")));
+
+        try (OutputStream os = storageService.write(privateLocation)) {
+            int _1mb = 1024 * 1024;
+            for (int i = 1; i <= loadS3TestFileSizeMb; i++) {
+                os.write(generateArrayWithSize(_1mb));
```
Comment (valb3r) [inline datasafe-storage/datasafe-storage-impl-s3/src/test/java/de/adorsys/datasafe/business/impl/storage/MultipartUploadS3SystemStorageServiceTest.java]: See de.adorsys.datasafe.business.impl.e2e.performance.services.ContentGenerator and its usages
```
try (OutputStream os = openWriteStream(user, oper)) {
            ByteStreams.copy(user.getGenerator().generate(oper.getContentId().getId()), os);
        }
``` 
it won't use any significant memory amount at all and you can move it to testshared

File: datasafe-storage/datasafe-storage-impl-s3/src/test/java/de/adorsys/datasafe/business/impl/storage/MultipartUploadS3SystemStorageServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                log.trace("Wrote {}mb of test file from {}mb", _1mb * i, loadS3TestFileSizeMb);
+            }
+        }
+
+        try(InputStream is = storageService.read(privateLocation)) {
+            String checksumOfFileFromS3 = checksum(is);
```
Comment (valb3r) [inline datasafe-storage/datasafe-storage-impl-s3/src/test/java/de/adorsys/datasafe/business/impl/storage/MultipartUploadS3SystemStorageServiceTest.java]: For checksum you can also use ContentGenerator:
digest(user.getGenerator().generate(oper.getExpected().getId()));

File: datasafe-storage/datasafe-storage-impl-s3/src/test/java/de/adorsys/datasafe/business/impl/storage/MultipartUploadS3SystemStorageServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            assertThat(testFileChecksum).isEqualTo(checksumOfFileFromS3);
+        } catch (IOException e) {
+            fail(e.getMessage(), e);
+        }
+    }
+
+    private void generateTestFile(String testFileName, int loadS3TestFileSizeMb) {
+        log.info("Starting write {} Mb file into {}", loadS3TestFileSizeMb, tempDir.toString());
+        try(FileOutputStream stream = new FileOutputStream(testFileName)) {
+            int _1mb = 1024 * 1024;
```
Comment (valb3r) [inline datasafe-storage/datasafe-storage-impl-s3/src/test/java/de/adorsys/datasafe/business/impl/storage/MultipartUploadS3SystemStorageServiceTest.java]: Please stick to naming convention - _1mb -> ONE_MB

File: datasafe-storage/datasafe-storage-impl-s3/src/test/java/de/adorsys/datasafe/business/impl/storage/MultipartUploadS3SystemStorageServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    private byte[] generateArrayWithSize(int _5mb) {
+        byte[] data = new byte[_5mb];
+
+        IntStream.range(0, _5mb).forEach(it -> data[it] = 'x');
+        return data;
+    }
+
+    @SneakyThrows
+    private String checksum(InputStream input) {
+        MessageDigest digest = MessageDigest.getInstance("SHA-256");
```
Comment (valb3r) [inline datasafe-storage/datasafe-storage-impl-s3/src/test/java/de/adorsys/datasafe/business/impl/storage/MultipartUploadS3SystemStorageServiceTest.java]: MD-5 is sufficient, it is fast enough

File: datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/business/impl/storage/MultipartS3StorageOutputStream.java [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+import java.io.IOException;
+import java.io.OutputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.*;
+
+@Slf4j
+public class MultipartS3StorageOutputStream extends OutputStream {
```
[the remark this answers is not in the document]
Comment (yatsenko-ihor) [reply datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/business/impl/storage/MultipartS3StorageOutputStream.java]: Replaced impl

File: datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java [hunk 6]
```diff
@@ -20,6 +21,26 @@
 import java.util.concurrent.ExecutorCompletionService;
 import java.util.concurrent.Executors;
 
+/**
```
Comment (valb3r) [inline datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java]: Move this to the top of the file

**Comment by valb3r** [comment]
Please fix failing build

File: datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+    private ByteArrayOutputStream currentOutputStream = new ByteArrayOutputStream(BUFFER_SIZE);
+
+    private InitiateMultipartUploadResult multiPartUploadResult;
+
+    private int partCounter = 1;
+
+    public MultipartUploadS3StorageOutputStream(String bucketName, ResourceLocation resource, AmazonS3 amazonS3) {
+        this.bucketName = bucketName;
+        this.objectName = resource.location().getPath().replaceFirst("^/", "");;
```
Comment (electronicpeter) [inline datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java]: This code is a copy paste of S3StorageService. Further it is not a good idea to choose the delimiter hard coded. The delimiter can be any character.

File: datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+    public void close() throws IOException {
+        synchronized (monitor) {
+            if (currentOutputStream == null) {
+                return;
+            }
+
+            if (isMultiPartUpload()) {
+                finishMultiPartUpload();
+
+                log.debug("Finished multi part upload");
```
Comment (electronicpeter) [inline datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java]: This logging should be done in finishMultiPartUpload, not here.

File: datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+            } else {
+                finishSimpleUpload();
+
+                log.debug("Finished simple upload");
```
Comment (electronicpeter) [inline datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java]: This logging should be done in finishMultiPartUpload, not here.

File: datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+
+        if (isMultiPartUpload()) {
+            amazonS3.abortMultipartUpload(new AbortMultipartUploadRequest(
+                            multiPartUploadResult.getBucketName(),
+                            multiPartUploadResult.getKey(),
+                            multiPartUploadResult.getUploadId()));
+        }
+    }
+
+    private List<PartETag> getMultiPartsUploadResults() throws ExecutionException, InterruptedException {
```
Comment (electronicpeter) [inline datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java]: We dont use Exceptions in the interface.

File: datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+    private int partCounter = 1;
+
+    public MultipartUploadS3StorageOutputStream(String bucketName, ResourceLocation resource, AmazonS3 amazonS3) {
+        this.bucketName = bucketName;
+        this.objectName = resource.location().getPath().replaceFirst("^/", "");;
+        this.amazonS3 = amazonS3;
+
+        log.debug("Write to bucket: {} with name: {}", Log.secure(bucketName), Log.secure(objectName));
+
+        completionService  = new ExecutorCompletionService<>(Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors()));
```
Comment (electronicpeter) [inline datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java]: Is here anything asynchron that requires opening own threads? Just asking.

File: datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java (hunk 7 above)
Comment (yatsenko-ihor) [reply datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java]: I think we have to replace it everywhere, not just here. Let's create new issue for that

File: datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java (hunk 10 above)
Comment (yatsenko-ihor) [reply datasafe-storage/datasafe-storage-impl-s3/src/main/java/de/adorsys/datasafe/storage/impl/s3/MultipartUploadS3StorageOutputStream.java]: this is private method which called in finishMultiPartUpload method. Please take a look finishMultiPartUpload method. Here we need exception. It's allow us handle interruption of uploading correctly

**Pull request merged**